[ao] Add histogram visualization capability to ModelReportVisualizer

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81975
* #81974
* #81973
* #81673

Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Add histogram visualization capability to ModelReportVisualizer

Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
Example of histogram
![thumbnail_IMG_1318](https://user-images.githubusercontent.com/35707033/180421161-f041d804-ca4d-4d61-920b-196ab890d2f0.jpeg)

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -526,22 +545,66 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+            return None
+
+        # get the x and y data and if per channel
+        x_data, y_data, data_per_channel = self._get_plottable_data(feature_filter, module_fqn_filter)
```
great, separating the code that process data and the code that visualizes data makes it cleaner

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -526,22 +545,66 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+
+        # for histogram, we just care about plotting the y data
+        # plot based on whether data is per channel or not
+        fig = plt.figure()
```
would be nice if we make this a separate function as well

vspenubarthi replied:
i think because some of the labels as well as legend, and titles are pretty different from histogram to line plot, it would be better to leave them separate for now.

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -526,22 +545,66 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
+        # checks if we have matplotlib and let's user know to install it if don't
+        if not got_matplotlib:
+            print("make sure to install matplotlib and try again.")
+            return None
```
Should we throw an exception here instead of silently failing? Otherwise the user may run the code and be confused why there's no visualization

vspenubarthi replied:
I had it like this earlier and it was failing some of the tests in github because the machines did not have matplotlib, and i wasn't sure how to install it.

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Add histogram visualization capability to ModelReportVisualizer"


Summary: This adds the capability to visualize the histogram in the
ModelReportVisualizer. You can visualize the histogram of a single
feature for a single layer, (for example you want to see the
distribution of some data across all channels), or for some feature
across multiple layers of a similar kind. All channel data is merged
together to plot one large distribution. The user gets to decide the
number of bins the histogram has and it will create those many equally
spaced bins.

    Expected Usage
    ```
    mod_rep_visualizer.generate_histogram_visualization(<feature_name>,<module_name>)
    ```

    You can also filter the modules so that only modules with a certain
    substring will have their features represented in the plot.

    > ** This is intended to be used in a `.ipynb` style notebook**

    The tests for this were just visual inspection for two reasons:
        1.) This method does not return anything, it just generates the
        visualization plot
        2.) All the data to create the plot visualization is gotten from
        `generate_filtered_tables` which is already tested, so testing all that
        for this again would be redundant.

    Example Image outputs are pasted below in the PR thread.

Test Plan: Visual Test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
